Move external interrupts into userspace

FawazTirmizi opened the pull request:
External interrupts are now handled in userspace, using the `riscv-plic` driver. The kernel now handles interrupts that go directly to the kernel, meaning that interrupt lines reserved for platform usage (`mcause` >= 16) are now usable in the kernel.

FawazTirmizi committed:
Added MutablePerfectHashMap and new functions

`MutablePerfectHashMap` is identical to `PerfectHashMap` except for the
fact that it has a new method: `set()`. This will write to the value
associated with the given key if there is already a value, otherwise
it will return an error.

This also adds the `contains()` method to both `PerfectHashMap` and
`MutablePerfectHashMap`. This method will indicate if there exists a
value for the given key.

FawazTirmizi committed:
Added a map of task name to ID to env var

Currently, it's impossible for an idol server to have a task slot for a
task which uses its client. As a result, the server's `build.rs` needs
to be able to turn task name strings into their ID number so that it can
send messages.

FawazTirmizi committed:
Add a RISC-V PLIC driver along with API.

Currently, external interrupts on rv32 targets are handled by the
kernel, and when a peripheral has an interrupt it's assumed that it's an
external one. This, however, makes it impossible for interrupt vectors
>= 16, which are reserved for platform use, to be used. Additionally,
there's no way to support platforms using another external interrupt
controller. By moving the PLIC to userspace, the kernel can properly
service interrupts to the other lines, and any interrupt controller can
be plugged in and used.

The API that has been added is intended to be general-purpose. Ideally,
this single API can be used for multiple interrupt controllers, allowing
for tasks to be portable across platforms implementing different
interrupt controllers.

FawazTirmizi committed:
Removed ext int handler and added platform handler

This now allows the kernel to handle interrupts coming in on lines
reserved for platform use, and having a task handle the external
interrupts.

FawazTirmizi committed:
Switched fe310-rtc to use external interrupt API

FawazTirmizi committed:
Added plic task to demo-hifive-inventor

beezow commented on build/xtask/src/config.rs:
```diff
@@ -189,6 +189,15 @@ impl Config {
             app_toml_path.to_str().unwrap().to_string(),
         );
 
+        let mut task_id_map: BTreeMap<String, u32> = BTreeMap::new();
+        for (i, (name, _)) in self.tasks.iter().enumerate() {
+            task_id_map.insert(name.to_string(), i.try_into().unwrap());
+        }
+        env.insert(
+            "HUBRIS_TASK_ID_MAP".to_string(),
+            ron::ser::to_string(&task_id_map).unwrap(),
+        );
```
This info allows you to create task that deadlock.  Can we add some comments/warning that misusing it may cause problems.

FawazTirmizi replied:
Done.

beezow commented on lib/phash/src/lib.rs:
```diff
@@ -34,6 +34,60 @@ impl<'a, K: Copy + PerfectHash + PartialEq, V> PerfectHashMap<'a, K, V> {
+    /// Looks up a value in the table by key, returning `None` if the key was
+    /// not stored in the table.
+    #[inline(always)]
+    pub fn get(&self, key: K) -> Option<&V> {
```
nit: From what I have seen, most maps in rust take a reference to the key `key: &K` not the key type itself.  Is that a style we should adopt?

FawazTirmizi replied:
I don't think so, the stuff that phash is currently used to hash is no bigger than 64 bits, and at that size it's probably more expensive to pass a reference. Plus that just causes us to diverge further from upstream. If we start phashing large objects then I think we should reevaluate.

beezow commented on drv/riscv-plic/src/main.rs:
```diff
@@ -0,0 +1,190 @@
+use idol_runtime::RequestError::Runtime;
+
+const PLIC_IRQ: u32 = 0x0000_0001;
+
```
`get_irq_owner` returns a `TaskId` while `get_task_irqs` takes a `InterruptOwner`.  Should these be using the same type?

beezow replied:
Figured out why as I was typing this.

furquan-rivos commented on lib/phash/src/lib.rs:
```diff
@@ -34,6 +34,60 @@ impl<'a, K: Copy + PerfectHash + PartialEq, V> PerfectHashMap<'a, K, V> {
+
+    #[inline(always)]
+    pub fn contains(&self, key: K) -> bool {
+        let i = key.phash(self.m) % self.values.len();
```
Shouldn't this also have a check for:
```
if self.values.is_empty() {
    return None
}
```

In fact, why not use the `get()` method itself?
```
pub fn contains(&self, key: K) -> bool {
      self.get(K).is_some()
}
```

FawazTirmizi replied:
You're right, I think this would panic on a call with an empty map. And I think using `get()` is fine.

furquan-rivos commented on drv/riscv-plic-api/src/lib.rs:
```diff
@@ -0,0 +1,20 @@
+    UnknownErr,
+}
+
+//pub fn wait_for_int(
```
Drop commented out code?

FawazTirmizi replied:
Missed that, thanks!

furquan-rivos commented on drv/riscv-plic/build.rs:
```diff
@@ -0,0 +1,163 @@
+        if task_set.insert(task.to_string()) == true {
+            writeln!(file, "task_slot!({}, {});", task.to_ascii_uppercase(), task);
+    }
+    */
```
drop commented out code?

FawazTirmizi replied:
Missed that too.

furquan-rivos commented on drv/riscv-plic/build.rs:
```diff
@@ -0,0 +1,163 @@
+                "(userlib::InterruptNum({}), (TaskId({}), 0b{:b})),",
+                irq.0, owner.0.0, owner.1
+            ),
+            None => "(userlib::InterruptNum::invalid(), userlib::InterruptOwner::invalid()),"
```
Shouldn't this be a build time failure instead of generating invalid number/owner?

FawazTirmizi replied:
This was lifted from `sys/kern/build.rs`, and I'm unsure why they had this as well. I'm guessing there is some situation where this can legally happen as they have other cases where the build will fail when making these maps. I can still have it fail instead if that's preferable.

furquan-rivos replied:
Yeah, I saw that. I was not sure why it was there. I think we can keep it as is for now (makes it easier to combine the code into a separate util crate).

furquan-rivos commented on drv/riscv-plic/build.rs:
```diff
@@ -0,0 +1,163 @@
+    pbits: u8,
+}
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
```
nit: I think a lot of this is extracted out of `sys/kern/build.rs`. Probably should be exposed as a build library that can be used by either kernel or any driver?

FawazTirmizi replied:
I agree. Because the maps can contain anything I'd expect there to be a good bit of weird templating involved, but I can see about adding this in a separate PR.

furquan-rivos replied:
SG.

furquan-rivos commented on drv/riscv-plic/build.rs:
```diff
@@ -0,0 +1,163 @@
+            notification: notification,
+        };
+        per_task_irqs.entry(owner).or_default().push(irq);
+    }
```
L100-L112: Do we really need to iterate over the TASK_CONFIG.tasks separately? Wouldn't it work just fine if you add `per_task_irqs.entry(owner).or_default().push(irq);` in the iterator above on lines 88-95?

FawazTirmizi replied:
I don't know how I missed that. Fixed.

furquan-rivos commented on drv/riscv-plic/build.rs:
```diff
@@ -0,0 +1,163 @@
+            task: task_id.index() as u32,
+            notification: notification,
+        };
+        per_task_irqs.entry(owner).or_default().push(irq);
```
Since `per_task_irqs` is building a vector of interrupt numbers, does it mean that a single task can request same notification for different interrupt numbers?

FawazTirmizi replied:
Yeah, there's nothing stopping a task from doing this.

furquan-rivos replied:
But in that case how would a task tell the plic driver which interrupt it really wants to mask/unmask/complete?

FawazTirmizi replied:
Presumably it'll do the operation on all of them; that's how it's done in the kernel (see the [irq_control](https://github.com/rivosinc/hubris/blob/502e5ae9797f846d65830742b0d6a99719513db7/sys/kern/src/syscalls.rs#L689-L719) syscall)

furquan-rivos commented on drv/riscv-plic/src/main.rs:
```diff
@@ -0,0 +1,190 @@
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                plic.mask(0, irqs[0].0.try_into().unwrap());
```
Is the assumption here that a task can have only a single interrupt number associated with a notification? This seems to contradict how the interrupt number is defined as a vector and handled in build.rs: https://github.com/rivosinc/hubris/pull/7/commits/95a67ca0eb0d3d2fb07d3aa1c51b1078439a5b52#diff-0d889df1d35cb9c1a214e153b69a2bc77b2fb16bf1f678b5eb35c54956a310dfR112

FawazTirmizi replied:
That's a mistake. Fixed.

furquan-rivos commented on drv/riscv-plic/src/main.rs:
```diff
@@ -0,0 +1,190 @@
+
+            let code = sys_post(owner.0, owner.1);
+
+            if code & 0xFFFF_FF00 == 0xFFFF_FF00 {
```
```
if code & FIRST_DEAD_CODE == FIRST_DEAD_CODE {
...
```

FawazTirmizi replied:
Did not realize that was defined already. Fixed.

furquan-rivos commented on drv/riscv-plic/src/main.rs:
```diff
@@ -0,0 +1,190 @@
+            if code & 0xFFFF_FF00 == 0xFFFF_FF00 {
+                let new_task_id = TaskId::for_index_and_gen(
+                    owner.0 .0.into(),
+                    ((code & 0x0000_00FF) as u8).into(),
```
~FIRST_DEAD_CODE?

FawazTirmizi replied:
Same as above. Fixed.

beezow commented on app/demo-hifive-inventor/app.toml:
```diff
@@ -6,25 +6,41 @@ stacksize = 896
+ints = [52]
+tasks = ['"rtc"']
+notification = [1]
+base = 0x0C00_0000
```
I am not a fan of having to define the base address for the PLIC both here and in the chip.toml.  Can this just be extracted from the chip.toml?

FawazTirmizi replied:
Will be fixed in next revision.

furquan-rivos commented on sys/kern/src/arch/riscv32.rs:
```diff
@@ -443,50 +444,43 @@ fn timer_handler() {
+        asm!("
+            csrw mie, {x}",
+            x = in(reg) new_mie,
+        );
```
Would `cur_mie.set(new_mie)` work instead of `asm`?

FawazTirmizi replied:
There is no general-purpose set function for `mie` in the `riscv` crate, for some reason.

furquan-rivos replied:
Aah. I probably misunderstood what `set` and `clear` mean in [mie.rs](https://github.com/rivosinc/riscv/blob/dev/fawaz/plic/src/register/mie.rs#L74)

FawazTirmizi replied:
They do generate functions for it, but they're private. That whole crate is wonky and needs a cleanup.

furquan-rivos commented on drv/fe310-rtc/src/main.rs:
```diff
@@ -5,19 +5,28 @@
     const RTC_INT: u32 = 0x1;
     ringbuf!(u64, 64, 0);
 
-    sys_irq_control(RTC_INT, true);
```
This change works fine for the current driver as it is RV specific. If we ever run into a case where a driver/task is shared by RV and ARM, then it could get really messy to handle the calls differently for each. Anyways, we don't have to worry about that now, but just something to keep in mind for the future.

FawazTirmizi replied:
I agree. I think it _is_ possible to make that work, but it would require some changes to the syscall internals that aren't worth doing until we run into this situation.

furquan-rivos replied:
Agreed. We can wait until we get there.

beezow commented on drv/fe310-rtc/src/main.rs:
```diff
@@ -5,19 +5,28 @@
+
+    match int_ctrl.disable_int(RTC_INT) {
+        Ok(()) => {}
+        Err(_) => sys_panic(b"RTC_INT is unassigned"),
```
Shouldn't all interrupts start off disabled?  Or is this just a "just in case" kinda thing?

FawazTirmizi replied:
Added as a just-in-case, and to ensure that I could test `enable_int()`.

beezow replied:
Fair enough.  Maybe add a comment to reflect this as the driver may be referenced on how to use the plic.

FawazTirmizi replied:
Done

furquan-rivos commented on app/demo-hifive-inventor/app.toml:
```diff
@@ -6,25 +6,41 @@ stacksize = 896
 max-sizes = { flash = 16384, ram = 2048 }
 start = true
-features = ["semihosting-riscv"]
+features = ["log-stringbuf"]
```
This is not related to plic. It should be a separate change outside of this series.

FawazTirmizi replied:
Agreed, will fix.

furquan-rivos commented on app/demo-hifive-inventor/app.toml:
```diff
@@ -6,25 +6,41 @@ stacksize = 896
+interrupts = { "plic.irq" = 0x1 }
+
+[tasks.ext_int_ctrl.config]
+ints = [52]
```
Another comment for the future. It would be good to see if we can extend the use of `.irq` to reference the interrupt numbers for peripherals. That will make it easier for app developers to just refer to that instead of putting magic numbers in here.

FawazTirmizi replied:
Yeah that definitely would be preferable. Doing that is going to require some more contortion of the build system that I decided isn't worth handling for now. But, there would be very little in this patch that needs to be changed to make that work so there isn't a major cost to switching.

furquan-rivos replied:
Agreed. I just wanted to note it. You can revisit that later.

furquan-rivos commented on app/demo-hifive-inventor/app.toml:
```diff
@@ -6,25 +6,41 @@ stacksize = 896
+ints = [52]
+tasks = ['"rtc"']
+notification = [1]
+base = 0x0C00_0000
```
Shouldn't this base come from a device "plic" in chip.toml?

FawazTirmizi replied:
The `chip.toml` isn't easily available to a task's `build.rs`, but I've already discussed this with @beezow and will be adding it in the next revision.

furquan-rivos replied:
Ack.

beezow commented on drv/fe310-rtc/src/main.rs:
```diff
@@ -5,19 +5,28 @@
+    let int_ctrl: RiscvIntCtrl =
+        RiscvIntCtrl::from(INT_CONTROLLER.get_task_id());
+
+    match int_ctrl.disable_int(RTC_INT) {
```
nit: I think this is equivalent to `.context("RTC_INT is unassignd").unwrap();`

However, I think`.unwrap()` could be good enough since the "unassignd" warning is already included in the error type

FawazTirmizi replied:
Changed.

furquan-rivos commented on drv/riscv-plic/src/main.rs:
```diff
@@ -0,0 +1,190 @@
+        loop {
+            let irq: u32 = match plic.claim(0) {
+                Some(irq) => core::primitive::u16::from(irq) as u32,
+                None => break,
```
Here and in the rest of the checks below, there is a break to exit out of the loop. What if there are more interrupts to claim? Shouldn't this be a continue instead of break?

FawazTirmizi replied:
The first break should be there as a claim returning 0 means there are no more interrupts. The latter two (lines 135 and 148) probably should be `continue` instead; those are cases where an unassigned interrupt has come in.

furquan-rivos replied:
Makes sense.

furquan-rivos commented on idl/riscv-plic.idol:
```diff
@@ -0,0 +1,35 @@
+Interface(
```
I think we should write some documentation regarding the use of PLIC driver by any task. There is certain expectation w.r.t. enable/disable and complete operations. It would be good to document this to make it easier for task developers.

FawazTirmizi replied:
Agreed. I plan on writing that once the API is approved.

beezow commented on app/demo-hifive-inventor/app.toml:
```diff
@@ -6,25 +6,41 @@ stacksize = 896
 
+[tasks.ext_int_ctrl]
+name = "drv-riscv-plic"
+priority = 1
```
Perhaps throw a warning that a task using the PLIC has a higher priority than the PLIC itself. This could cause starvation of interrupts if there is a third task in the mix.

beezow replied:
Turns out this already happens since the driver task has the PLIC in a task slot

beezow commented on drv/fe310-rtc/src/main.rs:
```diff
@@ -5,19 +5,28 @@
+
 #[export_name = "main"]
 fn main() -> ! {
     const RTC_INT: u32 = 0x1;
```
Would be nice if this was automatically exported based on the app.toml

FawazTirmizi replied:
The name of the task slot (`INT_CONTROLLER` here) is only defined within this file, so it's not feasible for now. But if there was some way to get it, this could be added in along with the const generation for peripheral addresses and notification masks.

beezow replied:
Missed the line with this one.  Should have been on 17. `const RTC_INT: u32 = 0x1;`

FawazTirmizi replied:
I'll be adding that in a later patch.

beezow commented on sys/kern/src/arch/riscv32.rs:
```diff
@@ -569,8 +560,22 @@ fn trap_handler(task: &mut task::Task) {
-            panic!("Unimplemented exception");
+            if mcause.is_exception() == true {
+                panic!("Unimplemented exception");
+            } else if (mcause.cause()
+                == Trap::Interrupt(Interrupt::MachineExternal))
+                || (mcause.code() >= 16)
```
Why isn't this just another match case like the rest of the exceptions?

beezow commented on sys/kern/src/arch/riscv32.rs:
```diff
@@ -569,8 +560,22 @@ fn trap_handler(task: &mut task::Task) {
-            panic!("Unimplemented exception");
+            if mcause.is_exception() == true {
+                panic!("Unimplemented exception");
+            } else if (mcause.cause()
+                == Trap::Interrupt(Interrupt::MachineExternal))
+                || (mcause.code() >= 16)
```
Is the `platform_interrupt_handler` supposed to catch all interrupts besides the timer?  So if the interrupt is not caught by this `if` then a task cannot  use it? 
If this is the case, I think the `16` should be a `12`, since anything `>=12` is a reserved interrupt.

beezow replied:
Along those same lines, is there a compile time check to ensure a task does not attempt to register an unsupported interrupt number?

FawazTirmizi replied:
The `platform_interrupt_handler` is supposed to handle the interrupts designated for platform use. Interrupts 12-15 are reserved by the ISA, so I think for now it's best to leave it as is. 

We don't have a compile-time check at the moment, but that's something that can be added, I'll put it in the kernel's `build.rs`.

beezow replied:
Where do you see interrupts >15 are for platform, while 12-15 are for ISA?
Under the mcause section, it has any interrupt >=12 is reserved.

beezow reviewed (changes requested):
I also tested this on the opentitan qemu machine with the watchdog. See [here](https://github.com/rivosinc/hubris/tree/dev/drew/opentitan).  It worked so I think baseline functionality seems good.

beezow commented on sys/kern/src/arch/riscv32.rs:
```diff
@@ -569,8 +560,22 @@ fn trap_handler(task: &mut task::Task) {
+                == Trap::Interrupt(Interrupt::MachineExternal))
+                || (mcause.code() >= 16)
+            {
+                // TODO: Put the second condition under a cfg
```
This `TODO` is pretty vague, Is this tracked in jira?

FawazTirmizi replied:
Meant to handle that in this patch, fixed.

FawazTirmizi commented on app/demo-hifive-inventor/app.toml:
```diff
@@ -6,25 +6,41 @@ stacksize = 896
+tasks = ['"rtc"']
+notification = [1]
+base = 0x0C00_0000
+pbits = 3
```
Because we don't actually set the priority to anything but 1 or 0, is there any benefit to having users set this?

beezow replied:
Can you just give it a default value?

FawazTirmizi replied:
I don't think so, the macro used to turn this into a struct for `build.rs` expects every field to be populated.

FawazTirmizi committed:
Export the path to the chip directory

This is exported as `HUBRIS_CHIP_DIR`. Doing this allows drivers to read
device info and generate constants for the device use. Eventually, the
build system should do this for you.

furquan-rivos commented on lib/phash/src/lib.rs:
```diff
@@ -34,6 +34,58 @@ impl<'a, K: Copy + PerfectHash + PartialEq, V> PerfectHashMap<'a, K, V> {
 
+    #[inline(always)]
+    pub fn contains(&self, key: K) -> bool {
+        return self.get(key).is_some();
```
nit: I think the typical way to write final return statement of a function/method in Rust is without `return` and `;`:

`self.get(key).is_some()`

FawazTirmizi replied:
Fixed.

furquan-rivos commented on build/xtask/src/config.rs:
```diff
@@ -180,10 +180,14 @@ impl Config {
 
+        // WARNING: Because this map provides task ID numbers directly, its
+        // contents can be used to send messages to tasks without creating a
+        // task_slot, so it's possible to deadlock. Use this with caution.
```
Just curious: Are there checks in other places currently to ensure that the tasks don't deadlock? Just trying to understand how adding to task_slot prevents deadlock.

FawazTirmizi replied:
The compiler won't let two tasks have task slots for each other, which is why this was needed in the first place. Otherwise, I could have done some weird stuff in the `build.rs` to generate all of the necessary slots.

furquan-rivos replied:
Aah I see. BTW was this comment supposed to be somewhere else. There is no map on line 186.

FawazTirmizi replied:
Yeah I think I squashed it into the wrong commit, so when I moved it I instead duplicated it.

beezow commented on drv/riscv-plic-server/src/main.rs:
```diff
@@ -0,0 +1,202 @@
+
+                // SAFETY: We already have the irq owner, so we know that this
+                // operation will succeed. No need to bother checking for errors.
+                unsafe {
```
Is there really a benefit to doing this unsafe, rather than just a normal unwrap?

FawazTirmizi replied:
The driver will have to check the result in case it's an error due to the irq not being assigned which is never the case as the irq would have been discarded in line 143. This prevents the need for that extra comparison.

beezow commented on drv/riscv-plic-server/src/main.rs:
```diff
@@ -0,0 +1,202 @@
+
+            let code = sys_post(owner.0, owner.1);
+
+            // The task that owns the line was restarted.
```
Should we disable a irq if the task that owns it was restarted?  
A task may have startup code that expects interrupts not yet be enabled.  How does the arm port handle this?

FawazTirmizi replied:
There's not any straightforward way to add this. We would have to somehow inform the interrupt controller that a task failed, which isn't possible unless the PLIC has a notification for said task, but at that point it's possible the task has already been restarted and is ready to receive interrupts again.

I don't think this is something we have to worry about, a task can just tell the controller to disable its' interrupts at the start of main().

furquan-rivos replied:
I think since this is a `sys_post` operation, it will only set the notification bits. So, unless the task asks the kernel for notifications, it won't really be interrupted.

FawazTirmizi replied:
Also, as far as I know the PLIC spec doesn't say that disabling an interrupt will actually disable it on the device, so this would require a lot more work than just figuring out `disable_int()` on a task's death.

beezow commented on sys/kern/src/arch/riscv32.rs:
```diff
@@ -443,50 +445,43 @@ fn timer_handler() {
 
-// Machine external interrupt handler
+// Handler for interrupts related to the platform
 #[no_mangle]
```
Why is this `#[no_mangle]`?

FawazTirmizi replied:
All interrupt handlers have `#[no_mangle]`, in both ARM and RISC-V. I'm not entirely certain why ARM does it, but I'm leaning towards keeping it for now.

beezow replied:
In arm they are referenced by the linker, but that is not the case on riscv right now.  I think it is fine to leave as is though.

beezow commented on sys/kern/src/arch/riscv32.rs:
```diff
@@ -569,8 +567,24 @@ fn trap_handler(task: &mut task::Task) {
         _ => {
-            panic!("Unimplemented exception");
+            cfg_if::cfg_if! {
+                if #[cfg(feature = "platform-interrupts")] {
```
nit: I think the naming scheme around the `platform-interrupts` feature and the `platform_interrupt_handler` is a tad confusing.  The feature flag implies that external interrupts are not platform interrupts, while the function name implies they are.
Isn't the `platform_interrupt_handler` more of a `userspace_interrupt_handler`?  It just delegates any interrupt to any task that has registered it.

FawazTirmizi replied:
It won't delegate any interrupt, only interrupts that are reserved for the platform. The feature flag is what enables tasks to request those irqs, hence the name.

beezow commented on drv/fe310-rtc/src/main.rs:
```diff
@@ -36,10 +46,10 @@ fn main() -> ! {
             ringbuf_entry!(num_ints);
 
-            sys_irq_control(RTC_INT, true);
+            unsafe { int_ctrl.complete_int(RTC_INT).unwrap_unchecked() };
```
Is the unsafe unwrap really necessary?

FawazTirmizi replied:
Because it's just a demo, no. I can change that.

furquan-rivos commented on drv/ext-int-ctrl-api/src/lib.rs:
```diff
@@ -0,0 +1,16 @@
+#[derive(Copy, Clone, Debug, FromPrimitive, IdolError)]
+pub enum ExtIntCtrlError {
+    IRQUnassigned,
+    UnknownErr,
```
This enum is unused.

FawazTirmizi replied:
Removed.

beezow reviewed (commented):
Looks pretty good, just some nitpick things.

furquan-rivos commented on drv/riscv-plic-server/build.rs:
```diff
@@ -0,0 +1,173 @@
+use std::path::PathBuf;
+
+task_config::task_config! {
+    ints: &'static [u16],
+    tasks: &'static [&'static str],
+    notification: &'static [u32],
```
I think this is fine. But another way to do this would be to put these three members together and use a vector of that structure. Something like this:

```
struct interrupt_config {
     int: u16,
     task: str,
     notification: u32,
}

task_config::task_config! {
      ints: Vec<interrupt_config>,
      pbits: u8,
}
```
It might things slightly easier to write in app.toml:
```
[tasks.ext_int_ctrl.config]
ints = {{int = 52, task = "rtc", notification = 1}, {int = 77, task = "dummy", notification = 1}}
pbits = 3
```

Anyways, I think we don't need to change anything right now. But if the app.toml starts getting confusing when adding more and more interrupts handled by plic, we can think about revisiting this.

FawazTirmizi replied:
I tried doing this, but the macro that parses this expects a very simple input. I can see if that can be changed later.

furquan-rivos replied:
Ah okay. SG.

furquan-rivos commented on drv/riscv-plic-server/build.rs:
```diff
@@ -0,0 +1,173 @@
+    pbits: u8,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+#[allow(dead_code)]
+struct Peripheral {
+    pub address: u32,
+    pub size: u32,
+    #[serde(default)]
+    pub interrupts: BTreeMap<String, u32>,
+}
```
This structure seems to be a copy of the `Peripheral` structure in build/xtask. I think we should avoid duplicating structures if possible. Maybe as a separate change this can be moved to a common place?

FawazTirmizi replied:
Agreed, I'll look into it for the future.

furquan-rivos commented on drv/fe310-rtc/src/main.rs:
```diff
@@ -5,19 +5,27 @@
 #![no_std]
```
Did you check if there is any difference in interrupt latency as seen by the task with and without this change?

FawazTirmizi replied:
There will definitely be an increase in interrupt latency, although we don't have any simple way to test it at the moment. There is an additional task switch that is needed as the driver will have to notify the task, and then another set of context switches so that the notified task gets control.

furquan-rivos replied:
Right. I think we should evaluate what that additional latency is. For some of our use cases, we might not be able to tolerate long latency. Can you please raise an issue to track this activity?

FawazTirmizi replied:
Agreed, and will do.

furquan-rivos commented:
LGTM. Ok to merge after @beezow's comments/concerns are addressed.
